File: sdk/cs/src/IModel.cs
// --------------------------------------------------------------------------------------------------------------------
// <copyright company="Microsoft">
//   Copyright (c) Microsoft. All rights reserved.
// </copyright>
// --------------------------------------------------------------------------------------------------------------------

namespace Microsoft.AI.Foundry.Local;

using System.Threading;
using System.Threading.Tasks;

public interface IModel
{
    string Id { get; }
    [System.Diagnostics.CodeAnalysis.SuppressMessage("Naming", "CA1716:Identifiers should not match keywords",
        Justification = "Alias is a suitable name in this context.")]
    string Alias { get; }

    ModelInfo Info { get; }

    Task<bool> IsCachedAsync(CancellationToken? ct = null);
    Task<bool> IsLoadedAsync(CancellationToken? ct = null);

    /// <summary>
    /// Download the model to local cache if not already present.
    /// </summary>
    /// <param name="downloadProgress">
    /// Optional progress callback for download progress.
    /// Percentage download (0 - 100.0) is reported.</param>
    /// <param name="ct">Optional cancellation token.</param>
    Task DownloadAsync(Action<float>? downloadProgress = null,
                       CancellationToken? ct = null);

    /// <summary>
    /// Gets the model path if cached.
    /// </summary>
    /// <param name="ct">Optional cancellation token.</param>
    /// <returns>Path of model directory.</returns>
    Task<string> GetPathAsync(CancellationToken? ct = null);

    /// <summary>
    /// Load the model into memory if not already loaded.
    /// </summary>
    /// <param name="ct">Optional cancellation token.</param>
    Task LoadAsync(CancellationToken? ct = null);

    /// <summary>
    /// Remove the model from the local cache.
    /// </summary>
    /// <param name="ct">Optional cancellation token.</param>
    Task RemoveFromCacheAsync(CancellationToken? ct = null);

    /// <summary>
    /// Unload the model if loaded.
    /// </summary>
    /// <param name="ct">Optional cancellation token.</param>
    Task UnloadAsync(CancellationToken? ct = null);

    /// <summary>
    /// Get an OpenAI API based ChatClient
    /// </summary>
    /// <param name="ct">Optional cancellation token.</param>
    /// <returns>OpenAI.ChatClient</returns>
    Task<OpenAIChatClient> GetChatClientAsync(CancellationToken? ct = null);

    /// <summary>
    /// Get an OpenAI API based AudioClient
    /// </summary>
    /// <param name="ct">Optional cancellation token.</param>
    /// <returns>OpenAI.AudioClient</returns>
    Task<OpenAIAudioClient> GetAudioClientAsync(CancellationToken? ct = null);

    /// <summary>
    /// Get an OpenAI API based EmbeddingClient
    /// </summary>
    /// <param name="ct">Optional cancellation token.</param>
    /// <returns>OpenAI.EmbeddingClient</returns>
    Task<OpenAIEmbeddingClient> GetEmbeddingClientAsync(CancellationToken? ct = null);

    /// <summary>
    /// Variants of the model that are available. Variants of the model are optimized for different devices.
    /// </summary>
    IReadOnlyList<IModel> Variants { get; }

    /// <summary>
    /// Select a model variant from <see cref="Variants"/> to use for <see cref="IModel"/> operations.
    /// An IModel from `Variants` can also be used directly.
    /// </summary>
    /// <param name="variant">Model variant to select. Must be one of the variants in <see cref="Variants"/>.</param>
    /// <exception cref="FoundryLocalException">If variant is not valid for this model.</exception>
    void SelectVariant(IModel variant);
}

File: sdk/python/src/imodel.py
# -------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.
# --------------------------------------------------------------------------
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Callable, List, Optional

from .openai.chat_client import ChatClient
from .openai.audio_client import AudioClient
from .openai.embedding_client import EmbeddingClient
from .detail.model_data_types import ModelInfo

class IModel(ABC):
    """Abstract interface for a model that can be downloaded, loaded, and used for inference."""

    @property
    @abstractmethod
    def id(self) -> str:
        """Unique model id."""
        pass

    @property
    @abstractmethod
    def alias(self) -> str:
        """Model alias."""
        pass

    @property
    @abstractmethod
    def info(self) -> ModelInfo:
        """Full model metadata."""
        pass

    @property
    @abstractmethod
    def is_cached(self) -> bool:
        """True if the model is present in the local cache."""
        pass

    @property
    @abstractmethod
    def is_loaded(self) -> bool:
        """True if the model is loaded into memory."""
        pass

    @property
    @abstractmethod
    def context_length(self) -> Optional[int]:
        """Maximum context length (in tokens) supported by the model, or ``None`` if unknown."""
        pass

    @property
    @abstractmethod
    def input_modalities(self) -> Optional[str]:
        """Comma-separated input modalities (e.g. ``"text,image"``), or ``None`` if unknown."""
        pass

    @property
    @abstractmethod
    def output_modalities(self) -> Optional[str]:
        """Comma-separated output modalities (e.g. ``"text"``), or ``None`` if unknown."""
        pass

    @property
    @abstractmethod
    def capabilities(self) -> Optional[str]:
        """Comma-separated capability tags (e.g. ``"chat,completion"``), or ``None`` if unknown."""
        pass

    @property
    @abstractmethod
    def supports_tool_calling(self) -> Optional[bool]:
        """Whether the model supports tool/function calling, or ``None`` if unknown."""
        pass

    @abstractmethod
    def download(self, progress_callback: Callable[[float], None] = None) -> None:
        """
        Download the model to local cache if not already present.
        :param progress_callback: Optional callback function for download progress as a percentage (0.0 to 100.0).
        """
        pass

    @abstractmethod
    def get_path(self) -> str:
        """
        Gets the model path if cached.
        :return: Path of model directory.
        """
        pass

    @abstractmethod
    def load(self) -> None:
        """
        Load the model into memory if not already loaded.
        """
        pass

    @abstractmethod
    def remove_from_cache(self) -> None:
        """
        Remove the model from the local cache.
        """
        pass

    @abstractmethod
    def unload(self) -> None:
        """
        Unload the model if loaded.
        """
        pass

    @abstractmethod
    def get_chat_client(self) -> ChatClient:
        """
        Get an OpenAI API based ChatClient.
        :return: ChatClient instance.
        """
        pass

    @abstractmethod
    def get_audio_client(self) -> AudioClient:
        """
        Get an OpenAI API based AudioClient.
        :return: AudioClient instance.
        """
        pass

    @abstractmethod
    def get_embedding_client(self) -> 'EmbeddingClient':
        """
        Get an OpenAI API based EmbeddingClient.
        :return: EmbeddingClient instance.
        """
        pass

    @property
    @abstractmethod
    def variants(self) -> List['IModel']:
        """Variants of the model that are available. Variants of the model are optimized for different devices."""
        pass

    @abstractmethod
    def select_variant(self, variant: 'IModel') -> None:
        """
        Select a model variant from ``variants`` to use for IModel operations.
        An IModel from ``variants`` can also be used directly.

        :param variant: Model variant to select. Must be one of the variants in ``variants``.
        :raises FoundryLocalException: If variant is not valid for this model.
        """
        pass
